File: linear list/array/Array.go
package src

import "fmt"

type array struct {
	items []interface{}
}

func NewArray(capacity int) *array {
	return &array{
		make([]interface{}, 0, capacity),
	}
}

// O(n)
func (a *array) Insert(item interface{}) {
	if len(a.items) == cap(a.items) {
		a.enlarge(cap(a.items) * 2)
	}
	a.items = append(a.items, item)
}

// O(n)
func (a *array) RemoveAt(index int) {
	if index < 0 || index > len(a.items) {
		panic("index error")
	}
	a.items = append(a.items[:index], a.items[index+1:]...)
}

// O(n)
func (a *array) IndexOf(element interface{}) int {
	for index, item := range a.items {
		if element == item {
			return index
		}
	}
	return -1
}

// O(n)
func (a *array) Print() {
	for _, item := range a.items {
		fmt.Printf("%v ", item)
	}
	fmt.Println("")
}

// O(n)
func (a *array) enlarge(capacity int) {
	newItems := make([]interface{}, cap(a.items), capacity)
	for index, item := range a.items {
		newItems[index] = item
	}
	a.items = newItems
}

File: linear list/array/Array.ts
class Array<T> {
  private items: T[];
  constructor(...items: T[]) {
    this.items = items;
  }
  // O(n)
  append(item: T) {
    this.items.push(item);
  }
  // O(n)
  removeAt(index: number) {
    if (index < 0 || index >= this.items.length)
      throw new Error("invalid index!");
    this.items.splice(index, 1);
  }
  // O(n)
  indexOf(item: T): number {
    for (let index = 0; index < this.items.length; index++)
      if (this.items[index] == item) return index;
    return -1;
  }
  // O(n)
  print() {
    for (let item of this.items) console.log(item);
  }
}

export default Array;
